library/ssl_tls13_server.c

```diff
@@ -335,6 +335,72 @@ static int ssl_tls13_check_ephemeral_key_exchange( mbedtls_ssl_context *ssl )
     return( 1 );
 }
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C) && \
+    defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+/*
+ * Try picking a certificate for this ciphersuite,
+ * return 0 on success and -1 on failure.
+ */
+static int ssl_tls13_pick_cert( mbedtls_ssl_context *ssl )
+{
+    mbedtls_ssl_key_cert *cur, *list;
+    const uint16_t *sig_alg = ssl->handshake->received_sig_algs;
+    uint16_t own_sig_alg;
+
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    if( ssl->handshake->sni_key_cert != NULL )
+        list = ssl->handshake->sni_key_cert;
+    else
+#endif /* MBEDTLS_SSL_SERVER_NAME_INDICATION */
+        list = ssl->conf->key_cert;
+
+    if( list == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "server has no certificate" ) );
+        return( -1 );
+    }
+
+    for( ; *sig_alg != MBEDTLS_TLS1_3_SIG_NONE; sig_alg++ )
+    {
+        for( cur = list; cur != NULL; cur = cur->next )
+        {
+            MBEDTLS_SSL_DEBUG_CRT( 3, "candidate certificate chain, certificate",
+                                   cur->cert );
+
+            /*
+            * This avoids sending the client a cert it'll reject based on
+            * keyUsage or other extensions.
+            */
+            if( mbedtls_x509_crt_check_key_usage(
+                        cur->cert, MBEDTLS_X509_KU_DIGITAL_SIGNATURE ) != 0 ||
+                mbedtls_x509_crt_check_extended_key_usage(
+                        cur->cert, MBEDTLS_OID_SERVER_AUTH,
+                        MBEDTLS_OID_SIZE( MBEDTLS_OID_SERVER_AUTH ) ) != 0 )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 3, ( "certificate mismatch: "
+                                      "(extended) key usage extension" ) );
+                continue;
+            }
+
+            if( ! mbedtls_ssl_tls13_get_sig_alg_from_pk(
```

Comment:
If the preferred algorithm is `MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA512`, the key size is 2048 and `MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA256` is a possible signature algorithm as well, I think that given the logic in `mbedtls_ssl_tls13_get_sig_alg_from_pk` we will end-up choosing `MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA256` instead of `MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA512`. This works fine for ECDSA but not RSA signature algorithms I'd say. By the way, in its usage in `ssl_tls13_write_certificate_verify_body`, there seems to be the same issue with `mbedtls_ssl_tls13_get_sig_alg_from_pk`: we may not end up using the preferred signature algorithm as we first choose the signature algorithm based on the key size. Thus overall, we probably need to change what we do in `mbedtls_ssl_tls13_get_sig_alg_from_pk` in the RSA case.

Reply:
This goes a bit beyond this PR. I will rework `mbedtls_ssl_tls13_get_sig_alg_from_pk` in a follow up PR. Let's forget about this in this PR.

Reply:
So many macro guards in the function mbedtls_ssl_tls13_get_sig_alg_from_pk(), MBEDTLS_RSA_C, MBEDTLS_PKCS1_V21, MBEDTLS_PKCS1_V15, MBEDTLS_SHA256_C, MBEDTLS_SHA384_C, MBEDTLS_SHA512_C. If we try to give out one concise function, seems the guards will be one obstacle.